src/csharp/Grpc.Core.Api/Metadata.cs

```diff
@@ -75,7 +75,67 @@ internal Metadata Freeze()
             return this;
         }
 
-        // TODO: add support for access by key
+        /// <summary>
+        /// Gets the last metadata entry with the specified key. If there are no matching entries then <c>null</c> is returned.
+        /// </summary>
+        public Entry Get(string key)
+        {
+            for (int i = entries.Count - 1; i >= 0; i--)
+            {
+                if (entries[i].Key == key)
+                {
+                    return entries[i];
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Gets the string value of the last metadata entry with the specified key. If there are no matching entries then <c>null</c> is returned.
+        /// </summary>
+        public string GetValue(string key)
+        {
+            return Get(key)?.Value;
+        }
+
+        /// <summary>
+        /// Gets the bytes value of the last metadata entry with the specified key. If there are no matching entries then <c>null</c> is returned.
+        /// </summary>
+        public byte[] GetValueBytes(string key)
```

Comment:
nit: add a note what happens if the key exists, but the value is not of type "bytes".